packages/ui/components/form/selectimproved/components/Select.tsx

```diff
@@ -0,0 +1,259 @@
+import SelectProvider from "./SelectProvider";
+import { Option } from "./type";
+
+interface SelectProps<T extends Option> {
+  options: T[];
+  selectedItems: T | T[] | null;
+  onChange: (value?: Option | Option[] | null) => void;
+  placeholder?: string;
+  isMultiple?: boolean;
+  isClearable?: boolean;
+  isSearchable?: boolean;
+  isDisabled?: boolean;
+  loading?: boolean;
+  menuIsOpen?: boolean;
+  searchInputPlaceholder?: string;
+  noOptionsMessage?: string;
+  classNames?: {
+    menuButton?: ({ isDisabled }: { isDisabled: boolean }) => string;
+    menu?: string;
+    tagItem?: ({ isDisabled }: { isDisabled: boolean }) => string;
+    tagItemText?: string;
+    tagItemIconContainer?: string;
+    tagItemIcon?: string;
+    list?: string;
+    listGroupLabel?: string;
+    listItem?: ({ isSelected }: { isSelected: boolean }) => string;
+    listDisabledItem?: string;
+    ChevronIcon?: ({ open }: { open: boolean }) => string;
+    searchContainer?: string;
+    searchBox?: string;
+    searchIcon?: string;
+    closeIcon?: string;
+  };
+}
+
+function Select<T extends Option>({
+  options = [],
+  selectedItems = null,
+  onChange,
+  placeholder = "Select...",
+  searchInputPlaceholder = "Search...",
+  isMultiple = false,
+  isClearable = false,
+  isSearchable = false,
+  isDisabled = false,
+  menuIsOpen = false,
+  classNames,
+}: SelectProps<T>) {
+  const [open, setOpen] = useState<boolean>(menuIsOpen);
+  const [inputValue, setInputValue] = useState<string>("");
+  const ref = useRef<HTMLDivElement>(null);
+  const searchBoxRef = useRef<HTMLInputElement>(null);
+  const escapePressed = useKeyPress("Escape");
+  const isMultipleValue = Array.isArray(selectedItems) && isMultiple;
+
+  const toggle = useCallback(() => {
+    if (!isDisabled) {
+      setOpen(!open);
+    }
+  }, [isDisabled, open]);
+
+  const closeDropDown = useCallback(() => {
+    if (open) setOpen(false);
+  }, [open]);
+
+  useOnClickOutside(ref, () => {
+    closeDropDown();
+  });
+
+  useEffect(() => {
+    if (escapePressed) {
+      closeDropDown();
+    }
+  }, [escapePressed, closeDropDown]);
+
+  const onPressEnterOrSpace = useCallback(
+    (e: React.KeyboardEvent<HTMLDivElement>) => {
+      e.preventDefault();
+      if ((e.code === "Enter" || e.code === "Space") && !isDisabled) {
+        toggle();
+      }
+    },
+    [isDisabled, toggle]
+  );
+
+  const removeItem = useCallback(
+    (item: Option) => {
+      // remove the item from the selected items
+      if (Array.isArray(selectedItems)) {
+        const newSelectedItems = selectedItems.filter((selectedItem) => selectedItem.value !== item.value);
+        onChange(newSelectedItems);
+      } else {
+        onChange(null);
+      }
+    },
+    [onChange, selectedItems]
+  );
+
+  const handleValueChange = useCallback(
+    (selected: Option) => {
+      function update() {
+        if (!isMultiple && !Array.isArray(selectedItems)) {
+          closeDropDown();
```

Comment:
Close when its a single array after selection

Reply:
Can you add this + a because as a comment? I struggle the see why it should be closed in that case 😁